### PR TITLE
Use map_partitions (Blockwise) in to_parquet

### LINKED ISSUE #8445: Suboptimal graph structure when read-writing a parquet
**What happened**:

to-parquet ops don't appear to "fuse" together with the rest of the linear chain. Since fusing chains is the current workaround for https://github.com/dask/distributed/issues/3974, this results in larger-than-necessary memory usage due to workers picking up work from other partitions earlier than they should.

**What you expected to happen**:

to-parquet ops should be merged into the linear chain like the other ops do. See code below for more explanation.

**Minimal Complete Verifiable Example**:

```python
import pandas as pd
import dask.dataframe as dd
import numpy as np

df = pd.DataFrame(np.random.randn(100,100), columns=[str(x) for x in range(100)])

ddf = dd.from_pandas(df, npartitions=2)
ddf.to_parquet("input")

ddf = dd.read_parquet("input")
ddf = ddf.to_parquet('test', compute=False)
ddf.visualize(optimize_graph=True)
```

This is the unoptimized graph:

Screen Shot 2021-12-02 at 3 23 53 PM

This is the optimized graph:

Screen Shot 2021-12-02 at 3 24 01 PM

Ideally the to-parquet op would be part of the fused chain. That way workers have no chance to prematurely switch to working on a different chain. I'm not aware of any reason that the to-parquet op should remain unfused.

**Anything else we need to know?**:

**Environment**:

- Dask version: 2021.11.2
- Python version: 3.8.x
- Operating System:
- Install method (conda, pip, source):

### PR DESCRIPTION
**Pull request by rjzamora**
~This PR adds a new `DataFrameOutputLayer` and uses it to move the data-writing component of `to_parquet` into a proper `HighLevelGraph` `Layer` class~ [**EDIT: This PR now uses `map_partitions` (with ~`partition_info`~ a new `BlockIndex(BlockwiseDep)` class) in lieu of a new `DataFrameOutputLayer`**].  Note that this PR required the column-projection changes from #8453, because the current behavior in `main` is actually a bug.

With these changes, the result of `ddf.visualize(optimize_graph=True)` (from [this](https://github.com/dask/dask/issues/8445#issue-1070070253) reproducer) is now:

Screen Shot 2021-12-14 at 4 07 35 PM

(The "read-parquet", "rename", "reset-index", and "to-parquet" tasks are all fused into "to-parquet")

- [x] Closes #8445
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [inline dask/dataframe/io/parquet/core.py:105]:
```suggestion
    Parquet Function-Wrapper Class
    
    Writes a DataFrame partition into a distinct
    parquet file. When called, the function also
    requires a ``partition_info`` kwarg.
```

**Comment by rjzamora** [comment]
Is anyone interested in reviewing this?  :D  (cc @jrbourbeau, @jsignell, @ian-r-rose, @gjoseph92)

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:98]: Why is the alias necessary?

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:133]:
```suggestion
    def __call__(self, df, *, partition_info):

        # Get partition index
        part_i = partition_info["number"]
```
I don't think `partition_info` is an optional argument. And `(partition_info or {}).get("number")` wouldn't work; the `part_i + self.i_offset` immediately after would fail if `part_i` was None.

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:149]:
```suggestion
               dict(self.kwargs_pass, head=True)
```
nit: Is `toolz.merge` faster? Otherwise, I lean slightly towards pure Python

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:788]: What if you implement `__dask_tokenize__` on `ToParquetFunctionWrapper`? Then I'd imagine the tokenization logic already in `map_partitions` would just work.

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:806]: Is that necessary? Doesn't seem like much of a tree, just a single reduction. Guess it's slightly more efficient?

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:130]: I like this use of #8310! However, the implementation of that is not ideal: we generate all the `partition_info` dicts up front, which is kinda overkill for large numbers of partitions.

For this case, could you use a `BlockwiseDep` instead that just returns the current block index? Something like https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L279-L288

With https://github.com/dask/dask/pull/7417, it seems like we're going with the pattern of having composable BlockwiseDep helpers as proposed in https://github.com/dask/dask/issues/7513. I _think_ (unconfirmed) you can even just pass them into `map_partitions` as arguments, but if not that would be a good change to make.

File: dask/dataframe/io/tests/test_parquet.py
Comment (gjoseph92) [inline dask/dataframe/io/tests/test_parquet.py:2464]:
```suggestion
    subgraph_rd = hlg_layer(dsk, "read-parquet")
```
https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/utils_test.py#L134-L139

File: dask/dataframe/io/tests/test_parquet.py
Comment (gjoseph92) [inline dask/dataframe/io/tests/test_parquet.py:2470]:
```suggestion
    subgraph_wt = hlg_layer(dsk, "to-parquet")
```

File: dask/layers.py
Comment (gjoseph92) [inline dask/layers.py:1220]: FYI, don't see this change in https://github.com/dask/dask/pull/8453

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:98]: We could just use `ParquetFunctionWrapper` and `ToParquetfunctionWrapper` and avoid this alias.  Does that sound reasonable?

Background: I just liked the "Read*" vs "To*" distinction in the names, and the only reason I used an alias after changing the "Read*" name is that `ParquetFunctionWrapper` is techincally a public API.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:806]: Why not? Right now its an all-to-one metadata reduction which we know is problematic at large scale.  If we use a formal HLG reduction layer, we will usually do the same thing, but will automatically get improved behavior at large scale.  Also, we will be removiong a "materialized layer" from the final HLG.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:133]: Ah , good point - I'm going to go with your [BlockwiseDep suggestion](https://github.com/dask/dask/pull/8487/files#r788329012), so this will need to change to `def __call__(self, df, partition_index):` anyway.

File: dask/dataframe/io/tests/test_parquet.py
Comment (rjzamora) [reply dask/dataframe/io/tests/test_parquet.py:2464]: cool!

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py]: Update: I removed the alias, so we just have `ParquetFunctionWrapper` and `ToParquetfunctionWrapper` for now.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py]: Good call - I added a `BlockIndex(BlockwiseDep)` class (similar to your suggestion here).

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [reply dask/dataframe/io/parquet/core.py:800]: Yeah, we should do that. My feeling was more that the tree-reduction layer doesn't get you many benefits at optimization compared to a low-level layer (no special fusion or anything), but I forgot that culling at least should be more efficient (when https://github.com/dask/dask/issues/8570 is resolved).

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:105]:
```suggestion
    requires the current block index (via ``blockwise.BlockIndex``).
```

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:128]:
```suggestion
    def __call__(self, df: pd.DataFrame, block_index: tuple[int]):
```

File: dask/blockwise.py
Comment (gjoseph92) [inline dask/blockwise.py:202]:
```suggestion
        # NOTE: unused, just needs to be set to follow the `BlockwiseDep` interface
        self.numblocks = numblocks
```
It confused me for a bit why this was necessary

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:128]: `df` may also be a `cudf.DataFrame` - Is there an established way that we are dealing with this ambiguity in Dask?

Comment (gjoseph92) [reply dask/dataframe/io/parquet/core.py:128]: ooof, probably not! maybe just leave this off then. the `block_index` one was more interesting to me anyway.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:783]: If we don't pass in `token=` explicitly, then `map_partitions` will produce a layer name of the form: `f"{funcname(func)}-{tokenize(func, meta, *args, **kwargs)}"`.  Therefore, we would also need to modify the `ToParquetFunctionWrapper` name to be `"to-parquet"`.  Is it worthwhile to add these defintions to `ToParquetFunctionWrapper` when we can just establish the name here?

Comment (gjoseph92) [reply dask/dataframe/io/parquet/core.py:783]: Wait, the `token=` argument to `map_partitions` is very much mis-named. It should really be called `name`:https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/dataframe/core.py#L6028-L6037

With what you have here, you'll get two tokens on the final name (`map_partitions` will append one automatically). I think just `token="to-parquet"` and a `__dask_tokenize__` method on `ToParquetFunctionWrapper` (and `BlockwiseDep`) is what you want here.

Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:783]: >Wait, the token= argument to map_partitions is very much mis-named. It should really be called name

Agree

>With what you have here, you'll get two tokens on the final name

Agree - I didn't realize this before, and I certainly don't like this.

>I think just token="to-parquet" and a __dask_tokenize__ method on ToParquetFunctionWrapper (and BlockwiseDep) is what you want here.

If we specify `token="to-parquet"`, then I don't think `ToParquetFunctionWrapper.__dask_tokenize__` will ever be used. In order to avoid making changes in `map_partitions`, I vote that we just define the `__repr__` of `ToParquetFunctionWrapper` to be `"to-parquet"`, and define `__dask_tokenize__` (as you suggested).

Comment (gjoseph92) [reply dask/dataframe/io/parquet/core.py]: > If we specify `token="to-parquet"`, then I don't think `ToParquetFunctionWrapper.__dask_tokenize__` will ever be used.

Don't think this is true: `ToParquetFunctionWrapper` will be part of `args` here, so it'll get tokenized as usual.

https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/dataframe/core.py#L6032-L6033

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:136]:
```suggestion
        return (
```
Minor note: I _think_ it's not actually necessary to tokenize all these things ourselves; you can just return the normalized form whose string repr you want to be md5'd
https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/base.py#L855

Doesn't matter at all though

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:133]: I'm surprised this worked. And it feels a little weird to me. I would have thought you'd just want to set `__name__` on the class?

https://github.com/dask/dask/blob/a504115312b1e591604276a3098b2e0076bf49e1/dask/utils.py#L852-L859

**Review by gjoseph92 — COMMENTED** [review]
Tiny nits, but I think this is ready, right? Just blocked by https://github.com/dask/dask/pull/8453?

**Comment by rjzamora** [comment]
>I think this is ready, right? Just blocked by #8453?

Thank for all the wonderful reviewing @gjoseph92 !  I hadn't really considered #8453 to be a blocker, since the primary purpose of that PR (in my mind) is the `creation_info` feature.  However, I now see that that PR includes changes that we have not really "agreed on," but have been copied into this PR.  Perhaps we should just agree on the pertinent changes here:

1. Does it make sense for `DataFrameLayer` to still exist now that it only acting as a label?
2. Do we really need to rease a `FutureWarning` when the "old" behavior of `project_columns` is encountered?

My interest in leaving `DataFrameLayer` around for now (1), is that I **do** intend to use it to store necessary information for multi-layer column projection (and evenutally predicate pushdown).  For example, we may use that class to define a base `required_columns` method designed to return the input columns required to produce a specific set of output columns (for that specific Layer).  This would be similar to `project_columns`, but we would be returning a set of column names, rahter than a new Layer.

Regarding (2): I have no problem with removing the deprecation warning and just letting old behavior fail. However, I seem to remember @ian-r-rose telling me that there is at least one `DataFrameIOLayer` definition in down-stream code.

**Comment by gjoseph92** [comment]
Personally, I'd prefer merging #8453 as a separate PR just to keep the git history cleaner. Nice to be able to refer back to the PR that introduced a change and figure out the intention for it.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [inline dask/dataframe/io/parquet/core.py:836]: We don't really "need" this to be a tree reduction (especially without defining the concat_func), but I felt it was worth making the minimal changes to move the entire `to_parquet` call to HLG.

Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py]: I've decided to roll back this change and save it for a distinct PR (focused only on parquet-metadata writing).

**Comment by rjzamora** [comment]
@martindurant - Do you know what's going on with the `test_parquet.py::test_timestamp96` CI failures? Is this maybe being cause by this PR (or my recent changes)?

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [inline dask/dataframe/io/parquet/core.py:819]:
```suggestion
    part_keys = df.__dask_keys__()
```
These aren't actually tasks, right? The name `parts_tasks` is confusing.

File: dask/dataframe/io/parquet/core.py
Comment (gjoseph92) [reply dask/dataframe/io/parquet/core.py]: Yeah, it's much harder to read as a tree reduction layer. Maybe a `tree_reduce` helper function for DataFrames would be worthwhile (`aca` could call into this too)?

**Comment by martindurant** [comment]
> Do you know what's going on with the test_parquet.py::test_timestamp96

pandas 1.4.0... #8626

File: dask/dataframe/io/parquet/arrow.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/arrow.py:766]: Yeah - It was just defined as a `classmethod` in `utils.py`, so I changed this to agree - but I guess it *could* be static if cls isn't used.

File: dask/dataframe/io/parquet/core.py
Comment (rjzamora) [reply dask/dataframe/io/parquet/core.py:819]: Right - We could use `data_write.__dask_keys__()` in place of `parts_tasks`

**Review by gjoseph92 — APPROVED** [review]
@rjzamora nice work, this seems good to me and will be a big improvement for users.

I did just realize we should probably open issues for converting the other `to_*` methods to use blockwise similarly.

**Comment by rjzamora** [comment]
Thanks for the reviews @gjoseph92 !

Just to avoid confusion - I'll probably wait to merge this until after the fastparquet failures are passing on main (and here).

**Comment by gjoseph92** [comment]
@rjzamora looks like tons of parquet tests failed only on windows 3.7. I'd assume this is all flakyness (workers couldn't start or something?), but slightly concerning to me that when that happened, the computation succeeded and just returned an unexpected result. I wonder if there's a race condition or something in those tests, and we could make them more resilient?

**Comment by rjzamora** [comment]
>looks like tons of parquet tests failed only on windows 3.7. I'd assume this is all flakyness (workers couldn't start or something?), but slightly concerning to me that when that happened, the computation succeeded and just returned an unexpected result. I wonder if there's a race condition or something in those tests, and we could make them more resilient?

Yeah - I haven't looked carefully through the windows 3.7 failures since many of those tests have been failing on all PRs for a while, but it does make sense to investigate ways to avoid flakyness and especially "silent" failures.

@gjoseph92 & @jrbourbeau - Is there any concern with merging this particular PR (given that the CI failures are the same here as elsewhere)?

**Comment by jsignell** [comment]
I am not concerned about merging this PR with failures that match those on main.

**Comment by rjzamora** [comment]
>I am not concerned about merging this PR with failures that match those on main.

Sounds good - I don't think anything is likely to change here, so I'll get it out of the way.
(It will also be nice to get task fusion in `to_parquet`!)

**Pull request merged**
